src/widget.cpp

```diff
@@ -694,10 +706,44 @@ void TextInputWidget::SetText(const std::string &text)
 	if (this->text_changed) this->text_changed();
 }
 
+/**
+ * Grab or release the keyboard focus.
+ * @param focus Widget should be focused.
+ */
+void TextInputWidget::SetFocus(bool focus)
+{
+	this->has_focus = focus;
+}
+
+bool TextInputWidget::OnClick([[maybe_unused]] const Point32 &base, const Point16 &pos)
+{
+	this->has_focus = true;
+
+	/* Find clicked cursor pos. \todo We might be able to use a more efficient logarithmic search. */
+	int w;
+	int text_len = this->buffer.size();
+	int best_x = INT_MAX;
+	for (int p = text_len; p >= 0; --p) {
+		_video.GetTextSize(this->buffer.substr(0, p), &w, nullptr, false);
```

Comment:
Depending on what magic the font renderer does, this might give wrong intermediate positions due to kerning.

Likely this doesn't affect us currently, as we paint each character separately? If so, another path for your 'todo' is to add character widths incrementally then. For current processors, that may even be the quickest way as they favor sequential access rather than jumping around (and then do sequential access for summing the widths) when you do bisection or so.